### PR TITLE
Fix Control Flow Graph - Grupo 3

### PR DESCRIPTION
**Pull request by delanoerhardt**
This is a WIP, the code is very messy, but most of the previously ignored tests now pass - with a new one being ignored.

This test suite seems quite weird:
- There are inconsistent tests, one includes the init statement for ForStmt, one doesn't.
- One of the tests create a graph where an assign statement circles back to itself, is that expected behaviour from assignments?
- Lots of tests using RepeatUntilStmt include the inner statement of the RepeatUntilStmt both in the repeat and in the list of statements used to generate the graph. This seems VERY wrong to me, the AST isn't generated like that and most of the the tests using ForStmts don't do that (another inconsistency - between For tests and loop tests in general).
- Some tests had requirements for node and edge amounts that didn't match the expected graph.

This is all to say that this test suite doesn't seem trustworthy. What that means is that I can't know whether I broke something or if the tests are broken themselves - or both - making this quite a painful part of the codebase to work with.

Complaining over, there's a quick list of changes made:
1. Added an unique id to Statements. This was needed as we must differentiate between equal - but not related - statements (i.e. the two `readInt(w);` in "stmt12"). Without a way to differentiate, every occurrence of an statement would be treated as a single node on the graph. **Maybe there's a better way to do this?**
2. Removed unused imports.
3. Created a new case class to hold the information of whether subnodes of an Statement A that should be included before A were inserted in the graph already.
4. Created a function to add edges respecting statements that need subnodes coming before themselves.

Regarding tests:
1. Tests stmt04.oberon and "Simple control flow graph with repeated statements" work after the changes;
2. Test stmt13.oberon was impossible to pass with the hardcode values in the asserts, maybe we shouldn't have those when we have the graphs to test against? The hardcoded asserts for node and edge amounts were commented out;
3. Tests stmt12.oberon, "RepeatUntil 05" and "RepeatUntilStmt 02" were very likely wrong, if I didn't miss anything;
4. Test "RepeatUntilStmt 03" broke after the changes but it seems to be wrong as well, I ignored it for now but as soon as I know what it is supposed to do I will fix it :).

### CODE REVIEW

File: src/main/scala/br/unb/cic/oberon/ast/OberonModule.scala [hunk 1]
```diff
@@ -172,6 +172,18 @@ case class AndExpression(left: Expression, right: Expression) extends Expression
...
+}
+
+object Statement {
+  private var nextId = 0
+  def getNextId() = {
+    val returnId = nextId;
+    nextId += 1;
+    returnId
+  }
 }
```
Comment (rbonifacio) [inline src/main/scala/br/unb/cic/oberon/ast/OberonModule.scala:187]: I liked this idea @delanoerhardt . Question: adding this `id` field did not break pattern matching, right?

File: src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala [hunk 2]
```diff
@@ -5,8 +5,18 @@ import scalax.collection.mutable.Graph
...
+}
+case object NoStmt extends StmtHolder[Nothing] {
+  def get: Nothing = throw new NoSuchElementException("NoStmt.get")
+}
+case class FullStmt(stmt: Statement) extends StmtHolder[Statement] {
+  def get: Statement = stmt
+}
+case class IgnoreInitStmt(stmt: Statement) extends StmtHolder[Statement] {
+  def get: Statement = stmt
+}
```
Comment (rbonifacio) [inline src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: could not understand the need for StmtHolder. Please, elaborate on that.

File: src/main/scala/br/unb/cic/oberon/ast/OberonModule.scala (hunk 1 above)
Comment (delanoerhardt) [reply src/main/scala/br/unb/cic/oberon/ast/OberonModule.scala:187]: It didn't, none of the other test suites were affected by this.

File: src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala (hunk 2 above)
Comment (delanoerhardt) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: ForStmt and RepeatUntilStmt need it, as they are loops that have sub-statements that go before themselves, I will call those pre-stmts.

To help illustrate why it is needed I will give an example, first without using it. Imagine the AST is composed of 3 statements, a simple statament 1, a ForStmt and another simple statement 2. The ForStmt has an Init Statement and an Inner Statement. We will assume Statements 1, 2, Init and Inner are simple just to make this explanation easier.

AST view

When IntraProceduralGraphBuilder tries to create an edge from Statement 1 to FromStmt using addEdge it will do the following: addEdge will receive Statement 1 as "from" and ForStmt as "target", then it will create an edge from Statement 1(from) to the Init Statement, then it will create an edge from the Init Statement to the ForStmt(target).

image

Indirect edge created, now we process Statement 1 to deal with its indirections. As it doesn't have any, we move on to creating an edge from ForStmt to Statement 2, quite simple.
 
image

But at this point ForStmt will be processed as well. This will eventually call processStmtNode with ForStmt as "from", Inner as "target" and ForStmt as "end" (this is a loop, after all). An edge will be created from ForStmt to Inner.

image

Now processStmtNode must also create an edge from "target" to "end", but how? If we simply use addEdge(target, end), Inner would be connected to Init, which is not intended.

image

We also could create the last edge by directly creating it from "target" to "end", as was done before. That would achieve the right result in this example, but the function processStmtNode is also used to deal with IfElseStmts, connecting an edge directly from an IfElseStmts to the "end" would create problems if "end" was a ForStmt:

image

How to solve this? I needed a way to differentiate whether the statement should be considered in full, including the pre-stmts beforehand, or if the pre-stmts should be disregarded.

Another way to achieve the right result in theses cases would be to **create a clone of processStmtNode to deal with loops specifically**. But if "target" is a StmtSequence processStmtNode delegates to createControlFlowGraph, so the problem persists and we would have to create a clone of createControlFlowGraph to deal with loops as well... seems to be a bad move.

tl;dr: What I ended up doing was: create a case class that will hold the information of WHERE the graph is supposed to go back to. I alse gave it a "get" function that mimics the behaviour of Option.get so that the cases where "get" is used don't change.

Is there a better solution to this? I guess I could create a simpler case class to encode the two valued cases and wrap it in an Option to get the NoStmt case as well. But I am not sure if that is better.

Comment (rbonifacio) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: I do not think this is necessary. I will take a look at the test cases to confirm that `for` and `repeat until` statements do require such special treatment. IMO, this part of the PR is adding too much unnecessary complexity.

Comment (rbonifacio) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: I understand your point. But there is a clever solution for building a control flow graph. I think that it is time to experiment with such an approach. Please, take a look at the book Principles of Program Analysis, sections 1.1 and 2.1. This book is freely available from Periódicos CAPES. 

Another option, which is our goal for this semester, is to convert every `repeat until`, `for` loop, and that like into `while` statements. 

Please, let me know which path are you planning to follow first.

Comment (rbonifacio) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: anyway, we already have a transformation that converts Oberon into Oberon Core.

Comment (rbonifacio) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: I can find some time tomorrow to improve the control flow graph, considering the recommendations from Principles of Program Analysis. Please, let me know if you prefer this alternative.

Comment (delanoerhardt) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: I think that would be great! Would you like me to create a new PR only with the addition of the Ids to statements? I think that alone fixes one of the tests (but it may break others, i haven't checked yet)

Comment (rbonifacio) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: Hi @delanoerhardt . I would love that. Could you please consider implementing a trait that has a label function (your Id) and then make Stmt mix with such a trait in the control flow graph module? In this way we could isolate this change better.

Many, many thanks for the discussion.

Comment (delanoerhardt) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: It seems like it is impossible to isolate the changes here.

We need to change the behaviour of either "==" comparison, hashCode or some other function that has a role in defining equality.

I tried using implicit classes, but it seems that those can't be used for overriding methods already defined, as is the case with the mentioned methods.

If you have any directions on how to do this I will look into it, otherwise this PR should be closed.

Comment (delanoerhardt) [reply src/main/scala/br/unb/cic/oberon/cfg/ControlFlowGraphBuilder.scala:20]: I thought of a way of fixing this just before sleeping lol.

Using implicit classes wasn't even needed, we have access to the SimpleNode class on this module.

I will create another PR with the fix
